# 1.Mapping, Annotation and QC

## Data Structure

```text
"shared_scripts"        # shared scripts by Lu Lab
"genomes/hg38"          # reference genomes (i.e. genome sequence and annotation)

github                  # my scripts
proj_exRNA
|-- RNAs_fasta          # rerference transcriptomes (fasta and index) 
|-- sample2_name             
`-- sample1_name        
    |-- fastq           # raw data: fastq files
    |-- fastqc          # QC of fastq
    |-- cutadapt        # trimmed fastq
    |-- collapsed       # de-duplicated fastq 
    `-- mapped          # mapped data: SAM/BAM files
```

### **Inputs**

| **File format** | **Information contained in file** | **File description** | **Notes** |
| --- | --- | --- |
| fastq | **reads** | small RNA-seq reads from single-end libraries. | This will be processed from long reads we got because it is cheaper \(see details in the protocol and example\). |
|  gtf | ** genome annotation** | Default genome annotation file is from GENCODE, mirBase, etc.  |  |

### **Outputs**

| **File format** | **Information contained in file** | **File description** | **Notes** |
| --- | --- | --- |
| bam | **alignments, i.e. mapped reads** | Produced by mapping reads to transcriptome | Raw reads are trimmed as single-ended, 50nt. We map to transcriptome for a better sensitivity \(see details in protocol and example\). |
| bt2 | **sequence** **index** | Indexed genome or transcriptomes | Different ncRNA types are indexed separately and mapped sequentially.  |

### Software/Scripts 

* bedtools
* fastqc
* cutadapt
* bowtie2

## Protocol

| Step | Input | Tool/script | Output | Working directory |
| :--- | :--- | :--- | :--- | :--- |
| 1.Preprocessing | 00.rawdata/\*.fastq | - | 02.rRNA/sampleID/sampleID.no\_rRNA.fastq |  |
| 1.1.index | genome/human\_hg38/gtf/\*.gtf | - | genome/human\_hg38/index/bowtie2\_index/\*.bt2 |  |
| 1.1.1 gtf-&gt;fasta | genome/human\_hg38/gtf/\*.gtf genome/human\_hg38/sequence/GRCh38.p12.genome.fa | bedtools | \*.fa |  |
| 1.1.2 fasta-&gt;index | \*.fa | bowtie2-build | \*.bt2 |  |
| 1.2 fastqc | 00.rawdata/\*.fastq | fastqc | 00.rawdata/\*\_fastqc.html | check raw reads' quality |
| 1.3 Remove 3' adapter and trim reads | 00.rawdata/\*.fastq | cutadapt | 01.cutAdapter/sampleID/sampleID.trimmed.cutAdapt3.fastq | trim low quality ends \(plus a hard cutoff: 50nt\) |
| 1.4 2nd fastqc | 01.cutAdapter/sampleID/sampleID.trimmed.cutAdapt3.fastq | fastqc | 01.cutAdapter/sampleID/sampleID.trimmed.cutAdapt3\_fastqc.html | make sure the low quality reads have been removed and/or trimmed |
| 1.5 Remove ribosomal RNA | 01.cutAdapter/sampleID/sampleID.trimmed.cutAdapt3.fastq + index file | bowtie2 | 02.rRNA/sampleID/sampleID.rRNA.sam 02.rRNA/sampleID/sampleID.no\_rRNA.fastq | - |
| 2.Mapping | 02.rRNA/sampleID/sampleID.no\_rRNA.fastq + index files | bowtie2 | 03.mapping/sampleID/01.miRNA/sampleID.miRNA.sam 03.mapping/sampleID/01.miRNA/sampleID.no\_miRNA.fastq --&gt;  03.mapping/sampleID/02.piRNA/sampleID.piRNA.sam 03.mapping/sampleID/02.piRNA/sampleID.no\_piRNA.fastq --&gt; ... | map to different RNA annotations step by step |



## Running Scripts

### Example of a single case

{% tabs %}
{% tab title="Set Parameters" %}
```bash
# genome seuqneces and annotaions
ln -s /BioII/lulab_b/shared/genomes ~/genomes
export hg38=~/genomes/human_hg38/sequence/GRCh38.p12.genome.fa
export gtf=~/genomes/human_hg38/gtf

# raw data (fastq files)
cd ~/proj_exRNA/example/fastq
ln -s /BioII/lulab_b/shared/projects/exRNA/hcc_examples/fastq/*.fastq .

# working space
cd ~/proj_exRNA
```
{% endtab %}

{% tab title="1.1 Index" %}
```bash
# Using miRNA as an example
# pwd: ~/proj_exRNA

#############################################################
# Setp 1.1: Index 
#############################################################
# 1.1.1: gtf2fasta - fetch sequeces from gtf annotated positions 
#############################################################

bedtools getfasta -fi $hg38 -bed $gtf/miRNA.gff -fo RNAs_fasta/miRNA.fa \
  > log.1.1.1 2>err.1.1.1 &
  
#############################################################
# 1.1.2: fasta2bt3 - index fasta file for bowtie 
#############################################################

bowtie2-build RNAs_fasta/miRNA.fa RNAs_fasta/miRNA \
  > log.1.1.2 2>err.1.1.2 &

```
{% endtab %}

{% tab title="1.2 QC & 1.3 trim" %}
```bash
#############################################################
# 1.2: fastaqc - fist round
#############################################################

fastqc -q -o example/fastqc miRNA.fastq 
#output the QC files to a dir examples/fastqc

#############################################################
# 1.3: trim step 1 - cut adaptor + remove last 100nt (keep 50nt) 
#############################################################
for i in `ls example/fastq/*.fastq`  ; do
  cutadapt -u -100 -q 30,30 --trim-n -m 15 \  
  -a AGATCGGAAGAGCACACGTCTGAACTCCAGTCAC -o $i.cutadapt  $i 
  # -u -100    remove last 100nt
  # -q 30,30   quality need to be above 30
  # -m 15      reads less than 15nt are removed
done
```
{% endtab %}

{% tab title="1.4 map and remove rRNA" %}
```bash
############################################################# 
# 1.4: map to rRNAs and keep the remaining reads for next 
#############################################################
```
{% endtab %}
{% endtabs %}

### [Script of batch job](https://github.com/urluzhi/scripts/blob/master/projects/exRNA/run_example.sh)

## Practice and Homework

Study and practice the followingï¼š[ ](https://youngleebbs.gitbooks.io/bioinformatics-training-program/content/exrna-seq-analysis/1preprocessing-mapping-and-qc.html)

* [Additional Tutorial](../getting-startted.md#learning-materials): 1.preprocessing\_mapping\_QC
* [Teaching PPT](../getting-startted.md#learning-materials): 
  * 0. Introduction to exRNA-seq; 
  * 1. Mapping, Annotation and QC
* [Teaching Video](../getting-startted.md#learning-materials):  
  * Week V - PARTII.1.Mapping etc - Wang
  * Week VI - PARTII. 2. run your job - example 1 bash  



