# 1.Mapping Annotation QC

## Software Requirement

* fastqc
* cutadapt
* bowtie2

## Data Requirement

```text
exRNA
|-- genome                # reference info (e.g. genome sequence and annotation)
`-- data
    |-- raw                 # raw data: fastq files
    |-- mapped              # mapped data: BAM files
    `-- de-duplicated       # extra fastq files for the de-duplication (collapse)
```

## Protocol

| Step | Input | Tool/script | Output | Working directory |
| :--- | :--- | :--- | :--- | :--- |
| 1.Preprocessing | 00.rawdata/\*.fastq | - | 02.rRNA/sampleID/sampleID.no\_rRNA.fastq |  |
| 1.1.index | genome/human\_hg38/gtf/\*.gtf | - | genome/human\_hg38/index/bowtie2\_index/\*.bt2 |  |
| 1.1.1 gtf-&gt;fasta | genome/human\_hg38/gtf/\*.gtf genome/human\_hg38/sequence/GRCh38.p12.genome.fa | bedtools | \*.fa |  |
| 1.1.2 fasta-&gt;index | \*.fa | bowtie2-build | \*.bt2 |  |
| 1.2 fastqc | 00.rawdata/\*.fastq | fastqc | 00.rawdata/\*\_fastqc.html | check raw reads' quality |
| 1.3 Remove 3' adapter and trim reads | 00.rawdata/\*.fastq | cutadapt | 01.cutAdapter/sampleID/sampleID.trimmed.cutAdapt3.fastq | trim low quality ends \(plus a hard cutoff: 50nt\) |
| 1.4 2nd fastqc | 01.cutAdapter/sampleID/sampleID.trimmed.cutAdapt3.fastq | fastqc | 01.cutAdapter/sampleID/sampleID.trimmed.cutAdapt3\_fastqc.html | make sure the low quality reads have been removed and/or trimmed |
| 1.5 Remove ribosomal RNA | 01.cutAdapter/sampleID/sampleID.trimmed.cutAdapt3.fastq + index file | bowtie2 | 02.rRNA/sampleID/sampleID.rRNA.sam 02.rRNA/sampleID/sampleID.no\_rRNA.fastq | - |
| 2.Mapping | 02.rRNA/sampleID/sampleID.no\_rRNA.fastq + index files | bowtie2 | 03.mapping/sampleID/01.miRNA/sampleID.miRNA.sam 03.mapping/sampleID/01.miRNA/sampleID.no\_miRNA.fastq --&gt;  03.mapping/sampleID/02.piRNA/sampleID.piRNA.sam 03.mapping/sampleID/02.piRNA/sampleID.no\_piRNA.fastq --&gt; ... | map to different RNA annotations step by step |



## Running Scripts

### Example of single case

{% tabs %}
{% tab title="Set Parameters" %}
```bash
export proj="~/projects/exRNA/example"
export hg38=~/genomes/human_hg38/sequence/GRCh38.p12.genome.fa
export gtf=~/genomes/human_hg38/gtf
```
{% endtab %}

{% tab title="1.1 Index" %}
```bash
#############################################################
# Setp 1.1: Index
#############################################################
# Step 1.1.1: gtf2fasta 
#############################################################
if [ "$1" = "1.1.1" ]; then
 for i in $RNAs ;do
  echo "start $i :"
 bedtools getfasta -fi $hg38 -bed $gtf/"$i".gff -fo RNAs_fasta/"$i".fa
 echo "$i finished."
 done
#############################################################
# Topic: 1.1.2: fasta2bt3 
#############################################################
elif [ "$1" = "1.1.2" ]; then
 for i in $RNAs ; do
  echo "start $i :"
 bowtie2-build RNAs_fasta/$i.fa RNAs_fasta/$i
 echo "$i finished."
 done




```
{% endtab %}

{% tab title="1.2 QC & 1.3 cutadapt" %}
```bash
#############################################################
# Topic: 1.2: fastaqc 
#############################################################
elif [ "$1" = "1.2" ]; then
 for i in `ls example/fastq/*.fastq`  ; do
  echo "start $i :"
 fastqc -q -o example/fastqc $i 
 echo "$i finished."
 done
#############################################################
# Topic: 1.3: cut adaptor 
#############################################################
elif [ "$1" = "1.3" ]; then
 for i in `ls example/fastq/*.fastq`  ; do
  echo "start $i :"
 cutadapt -u -100 -q 30,30 --trim-n -m 15 -a AGATCGGAAGAGCACACGTCTGAACTCCAGTCAC -o $i.cutadapt  $i 
 echo "$i finished."
 done
```
{% endtab %}
{% endtabs %}

### [Bash scripts for batch job](https://github.com/urluzhi/scripts/blob/master/projects/exRNA/run_example.sh)

## Practice and Homework

Study and practice the followingï¼š[ ](https://youngleebbs.gitbooks.io/bioinformatics-training-program/content/exrna-seq-analysis/1preprocessing-mapping-and-qc.html)

* [Additional Tutorial](../getting-startted.md#learning-materials): 1.preprocessing\_mapping\_QC
* [Teaching PPT](../getting-startted.md#learning-materials): 
  * 0. Introduction to exRNA-seq; 
  * 1. Mapping, Annotation and QC
* [Teaching Video](../getting-startted.md#learning-materials):  
  * Week V - PARTII.1.Mapping etc - Wang
  * Week VI - PARTII. 2. run 



