# 2.Expression Matrix

## Pipeline

![](../.gitbook/assets/expression.png)

## Data Structure

```text
"genomes/hg38/"         # reference genomes (i.e. genome sequence and annotation)
"shared_scripts/"       # shared scripts by Lu Lab
~/github                # I sync my own scripts to github

~/proj_exRNA/
|-- RNA_index           # rerference transcriptomes (fasta and index) 
|-- sample*_name        # ...
|-- sample2_name        # different samples     
`-- sample1_name        
    |-- fastq           # raw data: fastq files
    |-- fastqc          # QC of fastq
    |-- trim            # trimmed fastq (e.g. 3' adaptor cutted)
    |-- mapped          # mapped data: SAM/BAM files
    `-- exp             # expression of each gene/ncRNA
```

### **Inputs**

| **File format** | **Information contained in file** | **File description** | **Notes** |
| --- | --- |
| bam | **alignments** | Produced by mapping reads to the transcriptome. | Reads are trimmed using a proprietary version of cutAdapt. We map to transcriptome for a better sensitivity \(see details in protocol and example\). |

### **Outputs**

| **File format** | **Information contained in file** | **File description** | **Notes** |
| --- | --- | --- |
| bigWig | **signal** | Normalized RNA-seq signal | Signals are generated for transcriptome both the plus and minus strands and for unique reads and unique+multimapping reads.  |
| tsv | **gene \(ncRNA\) quantifications** | Non-normalized counts. |   |



## Running Scripts

### Software/Tools 

* RSEM
* homer
* HTseq
* FeatureCounts

### Example of single case

{% tabs %}
{% tab title="Set Parameters" %}
```bash
# genome seuqneces and annotaions
ln -s /BioII/lulab_b/shared/genomes ~/genomes
export hg38=~/genomes/human_hg38/sequence/GRCh38.p12.genome.fa
export gtf=~/genomes/human_hg38/gtf

# working space
cd ~/proj_exRNA
```
{% endtab %}

{% tab title="0.Convert & View" %}
```bash
# first convert relative positions to genomic coordinates
# then convert bam to bigWig, so that we can view it in genome browser

#####################################################################
#0.1 convert the relative positions on RNAs to genomic coordinates 
#####################################################################

rsem-tbam2gbam miRNA.indexDir/ NC_1.miRNA.sam NC_1.miRNA.rsem.bam

#####################################################################
#0.2 two ways to convert bam/sam to bedGraph/bigwig for visulaization 
#####################################################################
#0.2a Using homer
#Make tag directories for each sample (e.g. NC_1)
#works with sam or bam (samtools must be installed for bam) 
makeTagDirectory NC_1.miRNA.tagsDir/ NC_1.miRNA.sorted.bam

## Add "-strand separate" for strand-specific sequencing 
makeUCSCfile NC_1.miRNA.tagsDir/ -fragLength given -o auto
# (repeat for other tag directories)
```
{% endtab %}

{% tab title="a. homer" %}
```text
path0=/Share/home/younglee/projects/exRNA/hcc_lulab

## gene ##
# merge expression matrix for each RNA type
for i in `ls $path0/04.counts`;
do j=${i%.*};
echo $j;
for k in miRNA piRNA Y_RNA snRNA snoRNA srpRNA tRNA lncRNA mRNA;
do echo $k;
sed '1d' $path0/04.counts/${j}/sequentialMap/${j}.${k}.homer.ct | cut -f 9 | sed -e "1i lulab_${j}" | sed 's/-/_/g' > $path0/tmp/${j}.${k}.homer.ct.tmp
done;
done;

for k in miRNA piRNA Y_RNA snRNA snoRNA srpRNA tRNA lncRNA mRNA;
do echo $k;
sed '1d' $path0/04.counts/Normal-1/sequentialMap/Normal-1.${k}.homer.ct | cut -f 1 | sed -e "s/^/${k}_/g" | sed -e "1i geneID" > $path0/tmp/$k.homer.header
done;

for i in homer;
do echo $i;
for k in miRNA piRNA Y_RNA snRNA snoRNA srpRNA tRNA lncRNA mRNA;
do echo $k;
paste $path0/tmp/$k.$i.header $path0/tmp/*.${k}.$i.ct.tmp > $path0/05.matrix/hcc_lulab.sequentialMap.$i.$k.mx
sed -i "1s/......\t//1" $path0/05.matrix/hcc_lulab.sequentialMap.$i.$k.mx
done;
done;


## merge all RNA-type
for i in `ls $path0/04.counts`;
do j=${i%.*};
echo $j;
for k in miRNA piRNA Y_RNA snRNA snoRNA srpRNA tRNA lncRNA mRNA;
do echo $k;
sed '1d' $path0/04.counts/${j}/sequentialMap/${j}.${k}.homer.ct | cut -f 9 > $path0/tmp/${j}.${k}.homer.ct.tmp
done;
for m in homer;
do echo $m;
cat $path0/tmp/${j}.miRNA.${m}.ct.tmp $path0/tmp/${j}.piRNA.${m}.ct.tmp $path0/tmp/${j}.Y_RNA.${m}.ct.tmp $path0/tmp/${j}.snRNA.${m}.ct.tmp $path0/tmp/${j}.snoRNA.${m}.ct.tmp $path0/tmp/${j}.srpRNA.${m}.ct.tmp $path0/tmp/${j}.tRNA.${m}.ct.tmp $path0/tmp/${j}.lncRNA.${m}.ct.tmp $path0/tmp/${j}.mRNA.${m}.ct.tmp | sed -e "1i lulab_${j}" | sed 's/-/_/g' > $path0/tmp/${j}.merged.${m}.ct.tmp
cat $path0/tmp/miRNA.${m}.header $path0/tmp/piRNA.${m}.header $path0/tmp/Y_RNA.${m}.header $path0/tmp/snRNA.${m}.header $path0/tmp/snoRNA.${m}.header $path0/tmp/srpRNA.${m}.header $path0/tmp/tRNA.${m}.header $path0/tmp/lncRNA.${m}.header $path0/tmp/mRNA.${m}.header | grep -v "geneID" | sed -e "1i geneID" > $path0/tmp/merged.${m}.header
done;
done;

for m in homer;
do echo $m;
paste $path0/tmp/merged.${m}.header $path0/tmp/*.merged.${m}.ct.tmp > $path0/05.matrix/hcc_lulab.sequentialMap.${m}.merged.mx
sed -i "1s/......\t//1" $path0/05.matrix/hcc_lulab.sequentialMap.${m}.merged.mx
done;
```
{% endtab %}

{% tab title="b. HTseq" %}
```text
path0=/Share/home/younglee/projects/exRNA/hcc_lulab
## gene ##
# merge expression matrix for each RNA type
for i in `ls $path0/04.counts`;
do j=${i%.*};
echo $j;
for k in miRNA piRNA Y_RNA snRNA snoRNA srpRNA tRNA lncRNA mRNA;
do echo $k;
head -n -5 $path0/04.counts/${j}/sequentialMap/${j}.${k}.htseq.ct | cut -f 2 | sed -e "1i lulab_${j}" | sed 's/-/_/g' > $path0/tmp/${j}.${k}.htseq.ct.tmp
done;
done;

for k in miRNA piRNA Y_RNA snRNA snoRNA srpRNA tRNA lncRNA mRNA;
do echo $k;
head -n -5 $path0/04.counts/Normal-1/sequentialMap/Normal-1.${k}.htseq.ct | cut -f 1 | sed -e "s/^/${k}_/g" | sed -e "1i geneID" > $path0/tmp/$k.htseq.header
done;

for i in htseq;
do echo $i;
for k in miRNA piRNA Y_RNA snRNA snoRNA srpRNA tRNA lncRNA mRNA;
do echo $k;
paste $path0/tmp/$k.$i.header $path0/tmp/*.${k}.$i.ct.tmp > $path0/05.matrix/hcc_lulab.sequentialMap.$i.$k.mx
sed -i "1s/......\t//1" $path0/05.matrix/hcc_lulab.sequentialMap.$i.$k.mx
done;
done;


## merge all RNA-type
for i in `ls $path0/04.counts`;
do j=${i%.*};
echo $j;
for k in miRNA piRNA Y_RNA snRNA snoRNA srpRNA tRNA lncRNA mRNA;
do echo $k;
head -n -5 $path0/04.counts/${j}/sequentialMap/${j}.${k}.htseq.ct | cut -f 2 > $path0/tmp/${j}.${k}.htseq.ct.tmp
done;
for m in htseq;
do echo $m;
cat $path0/tmp/${j}.miRNA.${m}.ct.tmp $path0/tmp/${j}.piRNA.${m}.ct.tmp $path0/tmp/${j}.Y_RNA.${m}.ct.tmp $path0/tmp/${j}.snRNA.${m}.ct.tmp $path0/tmp/${j}.snoRNA.${m}.ct.tmp $path0/tmp/${j}.srpRNA.${m}.ct.tmp $path0/tmp/${j}.tRNA.${m}.ct.tmp $path0/tmp/${j}.lncRNA.${m}.ct.tmp $path0/tmp/${j}.mRNA.${m}.ct.tmp | sed -e "1i lulab_${j}" | sed 's/-/_/g' > $path0/tmp/${j}.merged.${m}.ct.tmp
cat $path0/tmp/miRNA.${m}.header $path0/tmp/piRNA.${m}.header $path0/tmp/Y_RNA.${m}.header $path0/tmp/snRNA.${m}.header $path0/tmp/snoRNA.${m}.header $path0/tmp/srpRNA.${m}.header $path0/tmp/tRNA.${m}.header $path0/tmp/lncRNA.${m}.header $path0/tmp/mRNA.${m}.header | grep -v "geneID" | sed -e "1i geneID" > $path0/tmp/merged.${m}.header
done;
done;

for m in htseq;
do echo $m;
paste $path0/tmp/merged.${m}.header $path0/tmp/*.merged.${m}.ct.tmp > $path0/05.matrix/hcc_lulab.sequentialMap.${m}.merged.mx
sed -i "1s/......\t//1" $path0/05.matrix/hcc_lulab.sequentialMap.${m}.merged.mx
done;
```
{% endtab %}

{% tab title="c.FeatureCounts" %}
```text
path0=/Share/home/younglee/projects/exRNA/hcc_lulab
## gene ##
# merge expression matrix for each RNA type
for i in `ls $path0/04.counts`;
do j=${i%.*};
echo $j;
for k in miRNA piRNA Y_RNA snRNA snoRNA srpRNA tRNA lncRNA mRNA;
do echo $k;
sed '1,2d' $path0/04.counts/${j}/sequentialMap/${j}.${k}.featureCounts.ct | cut -f 7 | sed -e "1i lulab_${j}" | sed 's/-/_/g' > $path0/tmp/${j}.${k}.featureCounts.ct.tmp
done;
done;

for k in miRNA piRNA Y_RNA snRNA snoRNA srpRNA tRNA lncRNA mRNA;
do echo $k;
sed '1,2d' $path0/04.counts/Normal-1/sequentialMap/Normal-1.${k}.featureCounts.ct | cut -f 1 | sed -e "s/^/${k}_/g" | sed -e "1i geneID" > $path0/tmp/$k.featureCounts.header
done;

for i in featureCounts;
do echo $i;
for k in miRNA piRNA Y_RNA snRNA snoRNA srpRNA tRNA lncRNA mRNA;
do echo $k;
paste $path0/tmp/$k.$i.header $path0/tmp/*.${k}.$i.ct.tmp > $path0/05.matrix/hcc_lulab.sequentialMap.$i.$k.mx
sed -i "1s/......\t//1" $path0/05.matrix/hcc_lulab.sequentialMap.$i.$k.mx
done;
done;


## merge all RNA-type
for i in `ls $path0/04.counts`;
do j=${i%.*};
echo $j;
for k in miRNA piRNA Y_RNA snRNA snoRNA srpRNA tRNA lncRNA mRNA;
do echo $k;
sed '1,2d' $path0/04.counts/${j}/sequentialMap/${j}.${k}.featureCounts.ct | cut -f 7 > $path0/tmp/${j}.${k}.featureCounts.ct.tmp
done;
for m in featureCounts;
do echo $m;
cat $path0/tmp/${j}.miRNA.${m}.ct.tmp $path0/tmp/${j}.piRNA.${m}.ct.tmp $path0/tmp/${j}.Y_RNA.${m}.ct.tmp $path0/tmp/${j}.snRNA.${m}.ct.tmp $path0/tmp/${j}.snoRNA.${m}.ct.tmp $path0/tmp/${j}.srpRNA.${m}.ct.tmp $path0/tmp/${j}.tRNA.${m}.ct.tmp $path0/tmp/${j}.lncRNA.${m}.ct.tmp $path0/tmp/${j}.mRNA.${m}.ct.tmp | sed -e "1i lulab_${j}" | sed 's/-/_/g' > $path0/tmp/${j}.merged.${m}.ct.tmp
cat $path0/tmp/miRNA.${m}.header $path0/tmp/piRNA.${m}.header $path0/tmp/Y_RNA.${m}.header $path0/tmp/snRNA.${m}.header $path0/tmp/snoRNA.${m}.header $path0/tmp/srpRNA.${m}.header $path0/tmp/tRNA.${m}.header $path0/tmp/lncRNA.${m}.header $path0/tmp/mRNA.${m}.header | grep -v "geneID" | sed -e "1i geneID" > $path0/tmp/merged.${m}.header
done;
done;

for m in featureCounts;
do echo $m;
paste $path0/tmp/merged.${m}.header $path0/tmp/*.merged.${m}.ct.tmp > $path0/05.matrix/hcc_lulab.sequentialMap.${m}.merged.mx
sed -i "1s/......\t//1" $path0/05.matrix/hcc_lulab.sequentialMap.${m}.merged.mx
done;
```
{% endtab %}
{% endtabs %}

[Example of batch job](https://github.com/lulab/training/tree/master/proj_exRNA/example_small)

## Tips/Utilities

#### Merge multiple bams to one 

```text
# mutiple bams to one bam file, in Snakemake form

rule mergeBam:
    input:
        bam = "02.mapping/{library}/sequentialMap/{library}.hg38other.bam"
    output:
        bam = "02.mapping/{library}/sequentialMap/{library}.sequentialMap.bam"
    params:
        jobname = "{library}.mergeBam",
        dir = "02.mapping/{library}/sequentialMap/"
    shell:
        """
        samtools merge {output.bam} {input.bam} `find {params.dir} -name "*.rsem.clean.bam"`

```

## Homework and more

1. Visualize your mapped reads with IGV \(locally\) and/or UCSC Genome Browser \(on line\).
2. Learn how to construct the expression matrix using HTSeq, featureCounts and homer; then compare the difference among these three methods. 

#### More Reading and Practice[ ](https://youngleebbs.gitbooks.io/bioinformatics-training-program/content/exrna-seq-analysis/1preprocessing-mapping-and-qc.html)

* [Additional Tutorial ](../getting-startted.md#learning-materials): 2. Construction of expression matrix



