# 2.Expression Matrix

## Data Structure

```text
exRNA
|-- genome                # reference info (e.g. genome sequence and annotation)
`-- data
    |-- raw                 # raw data: fastq files
    |-- mapped              # mapped data: BAM files
    `-- de-duplicated       # extra fastq files for the de-duplication (collapse)
```

### 

### **Inputs**

| **File format** | **Information contained in file** | **File description** | **Notes** |
| --- | --- | --- | --- |
| bam | **alignments** | Produced by mapping reads to the transcriptome. | Reads are trimmed using a proprietary version of cutAdapt. We map to transcriptome for a better sensitivity \(see details in protocol and example\). |
|  |  |  |  |
| tsv | **gene \(microRNA\) quantifications** | Non-normalized counts. |   |

### **Outputs**

| **File format** | **Information contained in file** | **File description** | **Notes** |
| --- | --- | --- |
| bigWig | **signal** | Normalized RNA-seq signal | Signals are generated for transcriptome both the plus and minus strands and for unique reads and unique+multimapping reads.  |
| tsv | **gene \(ncRNA\) quantifications** | Non-normalized counts. |   |



â€‹

| **Step** | **Input** | **Tool/script** | **Output** | **Notes** |
| :--- | :--- | :--- | :--- | :--- |
| convert mapped results | \*.bam |  | \*.bam | convert relative positions to genomic coordinates |
| 1.6 Convert and View mapped reads | \*.bam | homer, samtools, IGV genome browser | \*.bigWig, \*.bam | convert bam to bigWig, then view it in genome browser  |
| 1.7 Merge to one bam | \*.bam |  | \*.bam | merge multiple bam files into one; |

### 

{% tabs %}
{% tab title="Set Parameters" %}
```bash
# genome seuqneces and annotaions
ln -s /BioII/lulab_b/shared/genomes ~/genomes
export hg38=~/genomes/human_hg38/sequence/GRCh38.p12.genome.fa
export gtf=~/genomes/human_hg38/gtf

# working space
cd ~/proj_exRNA
```
{% endtab %}

{% tab title="0.Convert & View" %}
```bash
#####################################################################
#0.1 convert the relative positions on RNAs to genomic coordinates 
#####################################################################

rsem-tbam2gbam miRNA.indexDir/ NC_1.miRNA.sam NC_1.miRNA.rsem.bam

#####################################################################
#0.2 two ways to convert bam/sam to bedGraph/bigwig for visulaization 
#####################################################################
#0.2a Using homer
#Make tag directories for each sample (e.g. NC_1)
#works with sam or bam (samtools must be installed for bam) 
makeTagDirectory NC_1.miRNA.tagsDir/ NC_1.miRNA.sorted.bam

## Add "-strand separate" for strand-specific sequencing 
makeUCSCfile NC_1.miRNA.tagsDir/ -fragLength given -o auto
# (repeat for other tag directories)
```
{% endtab %}

{% tab title="a. homer" %}

{% endtab %}

{% tab title="b. HTseq" %}

{% endtab %}

{% tab title="c.FeatureCounts" %}

{% endtab %}
{% endtabs %}





## Practice and Homework

[Additional Tutorial ](../getting-startted.md#learning-materials): 2. Construction of expression matrix

