# 2.Expression Matrix

## Data Structure

```text
exRNA
|-- genome                # reference info (e.g. genome sequence and annotation)
`-- data
    |-- raw                 # raw data: fastq files
    |-- mapped              # mapped data: BAM files
    `-- de-duplicated       # extra fastq files for the de-duplication (collapse)
```

### 

### **Inputs**

| **File format** | **Information contained in file** | **File description** | **Notes** |
| --- | --- | --- | --- |
| bam | **alignments** | Produced by mapping reads to the transcriptome. | Reads are trimmed using a proprietary version of cutAdapt. We map to transcriptome for a better sensitivity \(see details in protocol and example\). |
|  |  |  |  |
| tsv | **gene \(microRNA\) quantifications** | Non-normalized counts. |   |

### **Outputs**

| **File format** | **Information contained in file** | **File description** | **Notes** |
| --- | --- | --- |
| bigWig | **signal** | Normalized RNA-seq signal | Signals are generated for transcriptome both the plus and minus strands and for unique reads and unique+multimapping reads.  |
| tsv | **gene \(ncRNA\) quantifications** | Non-normalized counts. |   |



## Practice and Homework

[Additional Tutorial ](../getting-startted.md#learning-materials): 2. Construction of expression matrix

