# 3.Differential Expression & Clustering

[Normalizing single-cell RNA sequencing data: challenges and opportunities, _Nature Methods_, 2017](https://www.ncbi.nlm.nih.gov/pubmed/28504683)

## Computational Methods of Differential Expression

**Assumption** for most normalization and differential expression analysis tools:

_The expression levels of most genes are similar, i.e., not differentially expressed._

**Recommendation for bulk RNA-seq**

* **RPM**: Read counts Per Million of total mapped reads; alternatives: RPKM, TPM
* **DEseq**: defines **scaling factor **\(also known as **size factor**\) estimates based on a pseudoreference sample, which is built with the geometric mean of gene counts across all cells \(samples\).
* [**TMM**](https://www.ncbi.nlm.nih.gov/pubmed/20196867)** **\(**EdgeR**\): trimmed mean of _M_ values

![](../.gitbook/assets/m.png)



**Performance**

![](../.gitbook/assets/performance.png)

### **Spike-in**

**RNA content \(total amount and species\) varies**

![](../.gitbook/assets/rna-content.png)

**for mRNAs:**

* 92 ERCC molecules
* 8 mRNAs
* whole transcriptome HeLa RNAs

**for sRNAs:**

* 52\(?\) sRNA sequences

**Caveat:**

Typically only half of the spike-in were detected.

