# 3.Differential Expression & Clustering

## Basic Methods of Differential Expression

**Assumption** for most normalization and differential expression analysis tools:

_The expression levels of most genes are similar, i.e., not differentially expressed._

**Recommendation for bulk RNA-seq**

* **RPM**: Read counts Per Million of total mapped reads; alternatives: RPKM, TPM
* **DEseq**: defines **scaling factor **\(also known as **size factor**\) estimates based on a pseudoreference sample, which is built with the geometric mean of gene counts across all cells \(samples\).
* [**TMM**](https://www.ncbi.nlm.nih.gov/pubmed/20196867)** **\(**EdgeR**\): trimmed mean of _M_ values

![](../.gitbook/assets/m.png)

**Performance**

![](../.gitbook/assets/performance.png)

## Conclusion

{% hint style="success" %}
**Take-Home Message:  **

The first three sections in PART II are the basic analyses. **For most users, you may stop here**. For advanced users, you can continue to the following sections. 
{% endhint %}

## More Readings and Practices

* **RNA-seq analysis** pipeline:: [https://github.com/mgonzalezporta/TeachingMaterial](https://github.com/mgonzalezporta/TeachingMaterial)
* **lncRNA analysis** pipeline: [http://webhome.weizmann.ac.il/home/igoru/PLAR/](http://webhome.weizmann.ac.il/home/igoru/PLAR/)
* **ENCODE** pipelines: [https://www.encodeproject.org/pipelines/](https://www.encodeproject.org/pipelines/)



