# 4.Normalization

## Background

### Problems and issues

* Sparsity of data and technical noise \("batch effects"\) --&gt; will mask the signal of interest

**Causes:**

![](../.gitbook/assets/noise.png)

### Spike-in for Normalization

**RNA content \(total amount and species\) varies**

![](../.gitbook/assets/rna-content.png)

**for mRNAs:**

* 92 ERCC molecules
* 8 mRNAs
* whole transcriptome HeLa RNAs

**for sRNAs:**

* 52\(?\) sRNA sequences

**Caveat:**

Typically only half of the spike-in were detected.

### Computational Normalization Tools

> **for Single cell RNA-seq \(and exRNA-seq\)**

1. **scran:**
   1. pools multiple cells \(samples\) in order to estimate cell-specific size factors in the **presence of zero inflation** and **unbalanced differential expression** of genes across groups of cells;
   2. **precluster** \(using e.g. rank-based clustering\) the cells into smaller, more homogeneous sets 
2. SCnorm
3. Census

If considering spike-ins:

1. SAMstrt
2. GRM

## **References**

* [Normalizing single-cell RNA sequencing data: challenges and opportunities, _Nature Methods_, 2017](https://www.ncbi.nlm.nih.gov/pubmed/28504683)

## **More Reading & Practice**

See more about **normalization**, **imputation** and **confounder** \(e.g. batch effect\) in

* [Additional Tutorial ](../getting-startted.md#learning-materials): 4.QC and Normalization; 5. Imputation and Confounders

## Video

### a\) Normalization 1

[@Youtube](https://youtu.be/QCxNIGrTWh0)

[@Bilibili](https://player.bilibili.com/player.html?aid=30592584&cid=53396844&page=1)

### b\) Normalization 2

[@Youtube](https://youtu.be/QDy2grY78DE)

[@Bilibili](https://player.bilibili.com/player.html?aid=30592643&cid=53396972&page=1)

